### ISSUE TITLE
Ability to plug in mock servers in `qx serve`/`qx test`

### ISSUE THREAD

**Issue by cboulanger** [qooxdoo/qooxdoo#10313]
Some of the tests in `qx.test.io.*` (like the GraphQL and Websockets implementations) rely on public external servers against which the protocol implementations are tested against. This is problematic since these servers might go away or change any time. For a more robust solution we should be able to plug in mock server classes into the built-in express server which are activated on demand (maybe using a compiler hint or a command line switch) and added as routes to the express server.

**Comment by hkollmann** [comment]
Why use mock? Just implement a local server in compile.js and run the test against this.

**Comment by cboulanger** [comment]
@cboulanger I am not sure if this works with `qx test`, but in any case you would have to implement a plug-in-like system just the same, since adding all of the different expected endpoints into one single `compile.js` creates a maintenance headache. I am thinking of a pluggable system where each test class can specify a corresponding server class which produces the expected response -- either statically or dynamically (probably needed for websockets, for example).

**Comment by cboulanger** [comment]
These server classes could have a `startUp()` and `tearDown()` method just like the client-side classes and turn the corresponding servers on and off on demand. So only those servers would be started that are actually needed by the test classes that are run client-side.

**Comment by hkollmann** [comment]
> @hkollmann I am not sure if this works with `qx test`, 

qx test has a needServer property which will start qx serve internally.

**Comment by hkollmann** [comment]
what's about a pluggable system based on qx serve and compile.js?

**Comment by cboulanger** [comment]
Sure, whatever works - my point is only that it should work on a per-test-class basis, so that only those server endpoints are activated which are actually needed by the test classes being run.

**Comment by cboulanger** [comment]
This needs a compile-time check which makes a list of the server classes that correspond to the `qx.test.*` classes that are being compiled and a process which runs the `startUp()` and `tearDown()` methods of these classes in `qx serve` and `qx test` before and after they do their main thing. I am not sure if that needs to happen in compile.js or just in the `qx serve` executable code, from which `qx test` inherits. This could be used outside of tests, actually, for any application that needs some additional server capabilities during development.

**Comment by cboulanger** [comment]
So maybe a compiler hint would be the most appropriate solution, such as `@serverClass` or something like that. So an application, for example, could add any type of node-based server by simply adding a server class and the compiler hint. What do you think, @johnspackman ?

**Comment by johnspackman** [comment]
I agree with @hkollmann that we don't need to mock, just implement the server - however, as @cboulanger points out there is coordination to be done between client and server for setup/teardown as well as collecting and collating the log output together.  It would be easier to develop the server as an app that is compiled and run instead of adding functionality to compile.js

IMHO the test system could benefit from a bit of an overhaul because it is generally client focused and we could better integrate the node tests as well as needing to add server tests.  The node tests we have a huge and messy proliferation of tests that are strapped together, some run as child processes, and when the github action fails it's a PITA to find which went wrong, and then it's often another PITA to just figure out which .js file to run and debug.  

Such a framework would be an excellent base for creating a demobrowser and a playground because they have the same needs.

**Comment by cboulanger** [comment]
Maybe the solution should be as general as possible so that tests would be just one use case. Wouldn't it be nice if the compiler, given a list of classes to compile, would generate a client-side and a server-side application simulataneously? It would have to know which class goes where, obviously...
